src/EventStore.ClientAPI/GossipSeed.cs

```diff
@@ -24,26 +25,36 @@ public class GossipSeed: IGossipSeed {
 		/// The host header to be sent when requesting gossip.
 		/// </summary>
 		public readonly string HostHeader;
+		
+		private readonly bool _v20Compatibility;
 
 		/// <summary>
 		/// Creates a new <see cref="GossipSeed" />.
 		/// </summary>
 		/// <param name="endPoint">The <see cref="IPEndPoint"/> for the External HTTP endpoint of the gossip seed. The standard port is 2113.</param>
 		/// <param name="hostHeader">The host header to be sent when requesting gossip. Defaults to String.Empty</param>
 		/// <param name="seedOverTls">Specifies that eventstore should use https when connecting to gossip</param>
-		public GossipSeed(IPEndPoint endPoint, string hostHeader = "", bool seedOverTls = false) {
+		/// <param name="v20Compatibility">If v20 compatibility mode is enabled.</param>
+		public GossipSeed(IPEndPoint endPoint, string hostHeader = "", bool seedOverTls = false, bool v20Compatibility = false) {
 			EndPoint = endPoint;
 			HostHeader = hostHeader == string.Empty ? null : hostHeader;
 			SeedOverTls = seedOverTls;
+			_v20Compatibility = v20Compatibility;
 		}
 		
 		public string ToHttpUrl() {
+			if (_v20Compatibility && !string.IsNullOrEmpty(HostHeader)) {
+				var scheme = SeedOverTls ? EndpointExtensions.HTTPS_SCHEMA : EndpointExtensions.HTTP_SCHEMA;
+
+				return string.Format("{0}://{1}:{2}/gossip?format=json", scheme, HostHeader, EndPoint.Port);
+			}
+			
 			return EndPoint.ToHttpUrl(SeedOverTls ? EndpointExtensions.HTTPS_SCHEMA : EndpointExtensions.HTTP_SCHEMA,
 				"/gossip?format=json");
 		}
 		
 		public string GetHostHeader() {
-			return HostHeader;
+			return _v20Compatibility ? HostHeader : "";
```

Comment:
@YoEight Isn't this a breaking change when _v20Compatibility=false? (for version 5 clients connecting against a version 5 server with a hostname configured)

Reply:
I believe that if we just return `HostHeader` here it should be ok.

The reason we had to clear the host header out for v20 client to connect against v5 server was due to the change in the DNS discovery mechanism in the v20 client - but it does not apply to IP gossip seeds as far as I can tell.

In this case, it seems that we don't even need the `_v20Compatibility` flag to be passed in the `GossipSeed` and no changes would need to be done to `GossipSeed.cs` at all.

Does this make sense or am I missing a case?

Reply:
While I agree now that we don't need to change `GetHostHeader` implementation for `GossipSeed`, We do need to send  that host even if we don't perform a DNS discovery in this case. The use-case is when we try to connect to a EventStoreDB Cloud cluster while using gossip seeds. You won't be able to send your gossip request without using the hostname (remember `SkipCertificateServer` is set to `false` by default)

Reply:
This is where the header is needed regardless if we do a DNS discovery or not: https://github.com/EventStore/EventStore/blob/6cf09ae8b144ad624f226c6fb784dc1592bacdae/src/EventStore.ClientAPI/Internal/ClusterDnsEndPointDiscoverer.cs#L230-L257